rust_src/src/multibyte.rs

```diff
@@ -42,32 +43,285 @@ use crate::{
+    }
+
+    // Equivalent to UNIBYTE_TO_CHAR
+    pub fn unibyte_to_char(self) -> Codepoint {
+        match self.is_ascii() {
+            true => self,
+            false => Codepoint::from_raw(self.0 as u8),
+        }
+    }
+
+    // Equivalent to MAKE_CHAR_MULTIBYTE
+    /// Transform an 8-bit codepoint to its byte8 encoded form.
+    pub fn to_multibyte(self) -> Codepoint {
+        debug_assert!(self.is_single_byte());
+        self.unibyte_to_char()
+    }
+
+    // Equivalent to CHAR_STRING
+    /// Write the codepoint to the given slice. The slice needs to be big
+    /// enough to hold the resulting bytes. Returns the amount of bytes written
+    pub fn write_to(self, to: &mut [u8]) -> usize {
+        let cp: u32 = self.into();
+        if cp <= MAX_1_BYTE_CHAR.into() {
+            to[0] = cp as u8;
+            1
+        } else if cp <= MAX_2_BYTE_CHAR {
+            // note: setting later bytes first to avoid multiple bound checks
+            to[1] = 0x80 | (cp & 0x3F) as u8;
+            to[0] = 0xC0 | (cp >> 6) as u8;
+            2
+        } else if cp <= MAX_3_BYTE_CHAR {
+            to[2] = 0x80 | (cp & 0x3F) as u8;
+            to[1] = 0x80 | ((cp >> 6) & 0x3F) as u8;
+            to[0] = 0xE0 | (cp >> 12) as u8;
+            3
+        } else if cp <= MAX_4_BYTE_CHAR {
+            to[3] = 0x80 | (cp & 0x3F) as u8;
+            to[2] = 0x80 | ((cp >> 6) & 0x3F) as u8;
+            to[1] = 0x80 | ((cp >> 12) & 0x3F) as u8;
+            to[0] = 0xF0 | (cp >> 18) as u8;
+            4
+        } else if cp <= MAX_5_BYTE_CHAR {
+            to[4] = 0x80 | (cp & 0x3F) as u8;
+            to[3] = 0x80 | ((cp >> 6) & 0x3F) as u8;
+            to[2] = 0x80 | ((cp >> 12) & 0x3F) as u8;
+            to[1] = 0x80 | ((cp >> 18) & 0x0F) as u8;
+            to[0] = 0xF8;
+            5
+        } else if cp <= MAX_CHAR {
+            let b = Codepoint::from(cp).to_byte8_unchecked();
+            to[1] = 0x80 | (b & 0x3F);
+            to[0] = 0xC0 | ((b >> 6) & 1);
+            2
+        } else {
+            error!("Invalid character: {:#x}", cp)
+        }
+    }
+
+    /// If character code C has modifier masks, reflect them to the character
+    /// code if possible. Return the resulting code.
+    pub fn resolve_modifier_mask(self) -> Codepoint {
+        let mut cp = self.0;
+        // A non-ASCII character can't reflect modifier bits to the code.
+        if !Codepoint::from(cp & !char_bits::CHAR_MODIFIER_MASK).is_ascii() {
+            return Codepoint::from(cp);
+        }
+        let ascii = (cp & 0x7F) as u8;
+        // For Meta, Shift, and Control modifiers, we need special care.
+        if cp & char_bits::CHAR_SHIFT != 0 {
+            let unshifted = cp & !char_bits::CHAR_SHIFT;
+            // Shift modifier is valid only with [A-Za-z].
+            // Shift modifier for control characters and SPC is ignored.
+            if (ascii >= b'A' && ascii <= b'Z') || ascii <= b' ' {
+                cp = unshifted;
```

Comment:
This logic is super confusing. But probably not more confusing than the original C.

Reply:
I'll try to see if I can refactor it to make it more readable, but I didn't write it, only made it a method, so I'm equally confused.

Reply:
Yea, I'm not going to make you rewrite it. I just can't tell if it's correct when looking at it. If the copies were faithful then it's all probably fine.